rasa/validator.py

```diff
@@ -483,3 +492,175 @@ def warn_if_config_mandatory_keys_are_not_set(self) -> None:
+            raise RasaException(
+                f"The slot '{constants.DIALOGUE_STACK_SLOT}' is used in the "
+                f"step '{step_id}' of flow id '{flow_id}', but it "
+                f"is a reserved slot. You must not use reserved slots in "
+                f"your flows.",
+            )
+
+    def verify_flows_steps_against_domain(self, user_flows: List[Flow]) -> bool:
+        """Checks flows steps' references against the domain file."""
+        all_good = True
+        domain_slots = {slot.name: slot for slot in self.domain.slots}
+        for flow in user_flows:
+            for step in flow.steps:
+                if isinstance(step, CollectInformationFlowStep):
+                    if step.collect_information not in domain_slots:
+                        raise RasaException(
+                            f"The slot '{step.collect_information}' is used in the "
+                            f"step '{step.id}' of flow id '{flow.id}', but it "
+                            f"is not listed in the domain slots. "
+                            f"You should add it to your domain file!",
+                        )
+                    current_slot = domain_slots[step.collect_information]
+                    self._raise_exception_if_list_slot(current_slot, step.id, flow.id)
+                    self._raise_exception_if_dialogue_stack_slot(
+                        current_slot, step.id, flow.id
+                    )
+
+                elif isinstance(step, SetSlotsFlowStep):
+                    for slot in step.slots:
+                        slot_name = slot["key"]
+                        if slot_name not in domain_slots:
+                            raise RasaException(
+                                f"The slot '{slot_name}' is used in the step "
+                                f"'{step.id}' of flow id '{flow.id}', but it "
+                                f"is not listed in the domain slots. "
+                                f"You should add it to your domain file!",
+                            )
+                        current_slot = domain_slots[slot_name]
+                        self._raise_exception_if_list_slot(
+                            current_slot, step.id, flow.id
+                        )
+                        self._raise_exception_if_dialogue_stack_slot(
+                            current_slot, step.id, flow.id
+                        )
+
+                elif isinstance(step, ActionFlowStep):
+                    if step.action not in self.domain.action_names_or_texts:
+                        raise RasaException(
+                            f"The action '{step.action}' is used in the step "
+                            f"'{step.id}' of flow id '{flow.id}', but it "
+                            f"is not listed in the domain file. "
+                            f"You should add it to your domain file!",
+                        )
+        return all_good
```

Comment:
Isn't this always returning `True`?

Reply:
Addressed: https://github.com/RasaHQ/rasa/pull/12866/commits/a1714b0e6c6e405a6a5f4bfde8419e8e93dc1a95